### PR TITLE
Add CBS unit tests and rewrite cbs.c

### PR DESCRIPTION
**Pull request by dcristoloveanu**
Add CBS unit tests and rewrite cbs.c

### CODE REVIEW

File: src/cbs.c [hunk 1]
```diff
@@ -7,153 +7,342 @@
...
-    if (key == NULL)
+    if (key_value == NULL)
     {
+        /* Codes_SRS_CBS_01_072: [ If constructing the message fails, `cbs_put_token_async` shall fail and return a non-zero value. ]*/
+        LogError("Failed creating value for property key %s", key);
         result = __FAILURE__;
     }
     else
     {
         AMQP_VALUE value_value = amqpvalue_create_string(value);
```
Comment (jebrando) [inline src/cbs.c]: >value_value [](start = 19, length = 11)

Kind of redundant.

File: src/cbs.c [hunk 2]
```diff
@@ -167,57 +356,102 @@ int cbs_close(CBS_HANDLE cbs)
...
+        /* Codes_SRS_CBS_01_048: [ `cbs_close` when not opened shall fail and return a non-zero value. ]*/
+        LogError("Already closed");
         result = __FAILURE__;
     }
     else
     {
+        if (cbs->cbs_state == CBS_STATE_OPENING)
+        {
+            /* Codes_SRS_CBS_01_079: [ If `cbs_close` is called while OPENING, the `on_cbs_open_complete` callback shall be called with `CBS_OPEN_CANCELLED`, while passing the `on_cbs_open_complete_context` as argument. ]*/
+            cbs->on_cbs_open_complete(cbs->on_cbs_open_complete_context, CBS_OPEN_CANCELLED);
```
Comment (jebrando) [inline src/cbs.c]: >cbs->on_cbs_open_complete(cbs->on_cbs_open_complete_context, CBS_OPEN_CANCELLED); [](start = 12, length = 81)

Is there anyway to actually cancel the underlying operation?  It seems that we're just cancelling in theory and not actually cancelling anything.

File: src/cbs.c [hunk 3]
```diff
@@ -167,57 +356,102 @@ int cbs_close(CBS_HANDLE cbs)
...
+        (token == NULL) ||
+        (on_cbs_put_token_complete == NULL))
     {
+        /* Codes_SRS_CBS_01_050: [ If any of the arguments `cbs`, `type`, `audience`, `token` or `on_cbs_put_token_complete` is NULL `cbs_put_token_async` shall fail and return a non-zero value. ]*/
+        LogError("Bad arguments: cbs = %p, type = %p, audience = %p, token = %p, on_cbs_put_token_complete = %p",
+            cbs, type, audience, token, on_cbs_put_token_complete);
+        result = __FAILURE__;
+    }
+    else if ((cbs->cbs_state == CBS_STATE_CLOSED) ||
+        (cbs->cbs_state == CBS_STATE_ERROR))
```
Comment (jebrando) [inline src/cbs.c]: >cbs->cbs_state [](start = 9, length = 14)

Would you need to do anything on opening?

File: src/cbs.c (hunk 2 above)
Comment (jebrando) [inline src/cbs.c]: Maybe put this after the close below.

---
In reply to: [107039232](https://github.com/Azure/azure-uamqp-c/pull/56#discussion_r107039232) [](ancestors = 107039232)

**Comment by jebrando** [comment]
:shipit:

File: src/cbs.c (hunk 1 above)
Comment (dcristoloveanu) [inline src/cbs.c]: Yeah. I guess. I'm going to leave it as such. I did consider changing it, but I got lazy between the CR and the IDE.

---
In reply to: [107037103](https://github.com/Azure/azure-uamqp-c/pull/56#discussion_r107037103) [](ancestors = 107037103)

File: src/cbs.c (hunk 2 above)
Comment (dcristoloveanu) [inline src/cbs.c]: That is a very interesting point. I shall move this below.

---
In reply to: [107039863](https://github.com/Azure/azure-uamqp-c/pull/56#discussion_r107039863) [](ancestors = 107039863,107039232)

Comment (dcristoloveanu) [inline src/cbs.c]: Done.

---
In reply to: [107044475](https://github.com/Azure/azure-uamqp-c/pull/56#discussion_r107044475) [](ancestors = 107044475,107039863,107039232)

File: src/cbs.c (hunk 3 above)
Comment (dcristoloveanu) [inline src/cbs.c]: You are so freaking smart!
Yes, I wanted to actually block put_token and delete_token in OPENING, but that would actually break IoTHub client and I would need lots of changes there to deal with it. WHich I did not quite want at this point...

---
In reply to: [107039495](https://github.com/Azure/azure-uamqp-c/pull/56#discussion_r107039495) [](ancestors = 107039495)

**Pull request merged**